src/shard_log.c

```diff
@@ -362,17 +396,26 @@ shard_log_iterator_init(cache              *cc,
    itor->cfg       = cfg;
 
    // traverse the log extents and calculate the required space
-   extent_addr = addr;
-   while (extent_addr != 0 && cache_get_ref(cc, extent_addr) > 0) {
+   bool page_terminal = FALSE;
+   extent_addr        = addr;
+   while (extent_addr != 0 && !page_terminal
+          && cache_get_ref(cc, extent_addr) > 0) {
       cache_prefetch(cc, extent_addr, PAGE_TYPE_FILTER);
       next_extent_addr = 0;
-      for (i = 0; i < pages_per_extent; i++) {
-         page_addr = extent_addr + i * shard_log_page_size(cfg);
-         page      = cache_get(cc, page_addr, TRUE, PAGE_TYPE_LOG);
-         if (shard_log_valid(cfg, page, magic)) {
-            num_valid_pages++;
-            itor->num_entries += ((shard_log_hdr *)page->data)->num_entries;
-            next_extent_addr = shard_log_next_extent_addr(cfg, page);
+      for (i = 0; i < pages_per_extent && !page_terminal; i++) {
+         page_addr          = extent_addr + i * shard_log_page_size(cfg);
+         page               = cache_get(cc, page_addr, TRUE, PAGE_TYPE_LOG);
+         shard_log_hdr *hdr = (shard_log_hdr *)page->data;
+         if (hdr->magic == magic) {
+            uint64 page_entry_count;
+            page_entry_count = num_log_entries(cfg, hdr, &page_terminal);
+            itor->num_entries += page_entry_count;
+            if (page_entry_count) {
+               num_valid_pages++;
+               next_extent_addr = shard_log_next_extent_addr(cfg, page);
+            }
+         } else {
+            page_terminal = TRUE;
```

Comment:
Relying on `hdr->magic` here seems like a problem. What if a log extent is deallocated and then reallocated as the next extent, but no writes have been issued to it yet. Then the magic will still match and all the checksums will work too.

Another issue is what happens with a torn writes that either occur at entry boundaries or which validate part of the header? It feels like we should still have a page-wide checksum.